src/app/reducers/references.handlers.ts

```diff
@@ -2,71 +2,87 @@ import { Node as ProsemirrorNode } from 'prosemirror-model';
 
 export function updateReference(state: ManuscriptHistoryState, payload: Reference): ManuscriptHistoryState {
   const referenceIndex = state.data.present.references.findIndex(({ id }) => id === payload.id);
+  const referenceChange = new UpdateObjectChange(
+    `references.${referenceIndex}`,
+    state.data.present.references[referenceIndex],
+    payload
+  );
+
   const { body } = state.data.present;
-  const changes = body.tr;
+  const transaction = body.tr;
   const newAttrs = { refId: payload.id, refText: getRefNodeText(payload) };
 
   body.doc.descendants((node: ProsemirrorNode, pos: number, parent: ProsemirrorNode) => {
     if (node.type.name === 'refCitation' && node.attrs['refId'] === payload.id) {
-      changes.replaceWith(pos, pos + node.nodeSize, body.schema.nodes['refCitation'].create(newAttrs));
+      transaction.replaceWith(pos, pos + node.nodeSize, body.schema.nodes['refCitation'].create(newAttrs));
     }
     return Boolean(node.childCount);
   });
-  const newState = updateManuscriptState(state.data, 'body', changes);
-  newState.present.references[referenceIndex] = payload;
-  sortReferencesList(newState.present.references);
-  newState.past[newState.past.length - 1]['references'] = state.data.present.references;
+
+  const change = new BatchChange([referenceChange, new ProsemirrorChange('body', transaction)]);
 
   return {
     ...state,
-    data: newState
+    data: {
+      past: [...state.data.past, change],
+      present: change.applyChange(state.data.present),
+      future: []
+    }
   };
 }
 
 export function addReference(state: ManuscriptHistoryState, payload: Reference): ManuscriptHistoryState {
-  const newDiff = createDiff({
-    references: state.data.present.references
-  });
+  const addReferenceChange = new AddObjectChange('references', payload, 'id');
+  const updatedManuscript = addReferenceChange.applyChange(state.data.present);
+  const updatedReferenceList = [...updatedManuscript.references];
+  sortReferencesList(updatedReferenceList);
 
-  const newManuscript = cloneManuscript(state.data.present);
-  newManuscript.references.push(payload);
-  sortReferencesList(newManuscript.references);
+  const sortReferencesChange = RearrangingChange.createFromListRearrange(
+    'references',
+    updatedManuscript.references,
+    updatedReferenceList
+  );
 
   return {
     ...state,
     data: {
-      past: [...state.data.past, newDiff],
-      present: newManuscript,
+      past: [...state.data.past, new BatchChange([addReferenceChange, sortReferencesChange])],
+      present: sortReferencesChange.applyChange(updatedManuscript),
       future: []
     }
   };
 }
 
 export function deleteReference(state: ManuscriptHistoryState, payload: Reference): ManuscriptHistoryState {
-  const referenceIndex = state.data.present.references.findIndex(({ id }) => id === payload.id);
-
+  const referenceChange = new DeleteObjectChange('references', payload, 'id');
```

Comment:
Similar to keywords, if I add a new reference, then delete it and use 'undo' to restore it. It deletes all the other references except for the one I restored which it has added twice.